packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts

```diff
@@ -0,0 +1,573 @@
+    }
+
+    if (!isPersistentKey(stateOrPersistentKey)) {
+      this.metrics?.stateClonedCount.observe(stateOrPersistentKey.clonedCount);
+      return stateOrPersistentKey;
+    }
+
+    return null;
+  }
+
+  /**
+   * Add a state of a checkpoint to this cache, prune from memory if necessary.
+   */
+  add(cp: phase0.Checkpoint, state: CachedBeaconStateAllForks): void {
+    const cpHex = toCheckpointHex(cp);
+    const key = toCheckpointKey(cpHex);
+    const stateOrPersistentKey = this.cache.get(key);
+    if (stateOrPersistentKey !== undefined) {
+      if (isPersistentKey(stateOrPersistentKey)) {
+        // was persisted to disk, set back to memory
+        this.cache.set(key, state);
+        this.persistentApis
+          .remove(stateOrPersistentKey)
+          .catch((e) =>
+            this.logger.debug(
+              "Error removing persisted state",
+              {key, stateOrPersistentKey: toHexString(stateOrPersistentKey)},
+              e
+            )
+          );
+        this.metrics?.stateRemoveCount.inc({reason: RemovePersistedStateReason.stateUpdate});
+      }
+      return;
+    }
+    this.metrics?.adds.inc();
+    this.cache.set(key, state);
+    this.epochIndex.getOrDefault(cp.epoch).add(cpHex.rootHex);
+    // don't prune from memory here, call it at the last 1/3 of slot 0 of an epoch
+  }
+
+  /**
+   * Searches in-memory state for the latest cached state with a `root` without reload, starting with `epoch` and descending
+   */
+  getLatest(rootHex: RootHex, maxEpoch: Epoch): CachedBeaconStateAllForks | null {
+    // sort epochs in descending order, only consider epochs lte `epoch`
+    const epochs = Array.from(this.epochIndex.keys())
+      .sort((a, b) => b - a)
+      .filter((e) => e <= maxEpoch);
+    for (const epoch of epochs) {
+      if (this.epochIndex.get(epoch)?.has(rootHex)) {
+        const inMemoryState = this.get({rootHex, epoch});
+        if (inMemoryState) {
+          return inMemoryState;
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Searches state for the latest cached state with a `root`, reload if needed, starting with `epoch` and descending
+   * This is expensive api, should only be called in some important flows:
+   * - Validate a gossip block
+   * - Get block for processing
+   * - Regen head state
+   */
+  async getOrReloadLatest(rootHex: RootHex, maxEpoch: Epoch): Promise<CachedBeaconStateAllForks | null> {
+    // sort epochs in descending order, only consider epochs lte `epoch`
+    const epochs = Array.from(this.epochIndex.keys())
+      .sort((a, b) => b - a)
+      .filter((e) => e <= maxEpoch);
+    for (const epoch of epochs) {
+      if (this.epochIndex.get(epoch)?.has(rootHex)) {
+        try {
+          const state = await this.getOrReload({rootHex, epoch});
+          if (state) {
+            return state;
+          }
+        } catch (e) {
+          this.logger.debug("Error get or reload state", {epoch, rootHex}, e as Error);
```

Comment:
getOrReload already swallows all errors, I need to look more into the usage pattern of this but on a cursory look error handling should be more explicit than converting all errors to null

Reply:
in `regen.get*()` apis, they only expect checkpoint state caches returning state or `null`. We could shift the error handling there but I think regen should also have to try/catch and try to load another checkpoint state. It depends on how we decide the api spec in not only this implementation but the generalized `CheckpointStateCache` interface and all of the implementations so that we have no issue switching the old `CheckpointStateCache` vs this new `PersistentCheckpointStateCache`